### ISSUE TITLE
Issues with http.keep-alive-timeout 0

### ISSUE THREAD

**Issue by zickgraf** [arangodb/arangodb#9288]
## My Environment

The default environment when running
```
docker run -e ARANGO_ROOT_PASSWORD=openSesame arangodb/arangodb:3.4.6.1
```

* __Infrastructure__:               own
* __Operating System__:        Arch Linux with kernel version 5.0.4
* __Total RAM in your machine__:        16GB
* __Disks in use__:        HDD

## Component, Query & Data

__Affected feature__:

When starting ArangoDB with `http.keep-alive-timeout 0`, trying to connect with `arangosh` fails.

## Steps to reproduce

1. Execute `docker run -e ARANGO_ROOT_PASSWORD=openSesame arangodb/arangodb:3.4.6.1 --http.keep-alive-timeout 0`
2. Get a shell with `docker exec -i -t $CONTAINER_ID /bin/sh`
3. Execute `arangosh --server.password openSesame`

__Problem__:
One of the following errors occur:
```

                                       _     
  __ _ _ __ __ _ _ __   __ _  ___  ___| |__  
 / _` | '__/ _` | '_ \ / _` |/ _ \/ __| '_ \ 
| (_| | | | (_| | | | | (_| | (_) \__ \ | | |
 \__,_|_|  \__,_|_| |_|\__, |\___/|___/_| |_|
                       |___/                 

arangosh (ArangoDB 3.4.6-1 [linux] 64bit, using jemalloc, build tags/v3.4.6.1-0-gd3c504bfc9, VPack 0.1.33, RocksDB 5.16.0, ICU 58.1, V8 5.7.492.77, OpenSSL 1.1.0j  20 Nov 2018)
Copyright (c) ArangoDB GmbH

Command-line history will be persisted when the shell is exited.
Could not connect to endpoint 'tcp://127.0.0.1:8529', database: '_system', username: 'root'

127.0.0.1:8529@_system>
```

```
Segmentation fault (core dumped)
```

```
Illegal instruction (core dumped)
```

Sometimes it also just blocks without any output.

__Expected result__:

`arangosh` successfully connects to the local server.

__Note__: we discovered this issue while trying to connect to an ArangoDB server behind Apache via reverse proxy. The Apache server has a keep alive timeout of 5 seconds. In this case, `arangosh` successfully connects but if one waits 5 seconds between entering two commands, the following error occurs:

```
JavaScript exception in file '/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js' at 99,7: ArangoError 2001: Connection closed by remote
!      throw error;
!      ^
stacktrace: ArangoError: Connection closed by remote
    at Object.exports.checkRequestResult (/usr/share/arangodb3/js/client/modules/@arangodb/arangosh.js:97:21)
    at Proxy.ArangoDatabase._collections (/usr/share/arangodb3/js/client/modules/@arangodb/arango-database.js:273:12)
    at <shell command>:1:4
```
Somtimes instead of `Connection closed by remote`, the error is `Connection reset by peer`. This might be connected.

**Comment by dothebart** [comment]
Hi, 
please note that arangodb needs to have access to the cpuid-feature in order to determine the available instructions. Whats your actual `cat /proc/cpuinfo` ?

**Comment by zickgraf** [comment]
The content of my `/proc/cpuinfo` is:
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 158
model name	: Intel(R) Core(TM) i5-7500T CPU @ 2.70GHz
stepping	: 9
microcode	: 0xb4
cpu MHz		: 897.631
cache size	: 6144 KB
physical id	: 0
siblings	: 4
core id		: 0
cpu cores	: 4
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 22
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp flush_l1d
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf
bogomips	: 5426.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 39 bits physical, 48 bits virtual
power management:
```
(repeated for each of the four cores)

One addition: without setting `--http.keep-alive-timeout 0` or with `--http.keep-alive-timeout 1`, everything works as expected.

**Comment by graetzer** [comment]
if you can access the arangodb instance without a proxy you can use `--server.endpoint vst://<ip>:<port>` as a workaround to keep accessing the server

**Comment by dothebart** [comment]
Ok, while this is obviously a bug in arangosh (which we have been able to reproduce) whats your reason to disable HTTP-Keepalive in firstplace? This sounds like a bad idea and the path to very bad performance...

**Comment by zickgraf** [comment]
As mentioned above, the goal is to run ArangoDB behind Apache via reverse proxy, but this leads to `Connection closed by remote` errors. As a workaround, I tried to disable Apache's keep alive feature, but this led to the errors reported in this issue. Since I could reproduce this without Apache, my hope was that (as a first step) this is easier to debug and fix and maybe even fixes the problem with Apache. If you want, I can add more details about the problem with Apache here or open a separate issue.

**Comment by graetzer** [comment]
seems to work now, more or less

**Comment by zickgraf** [comment]
I can confirm that the segmentation faults no longer happen in version 3.5.2. However, I still get
```
Could not connect to endpoint 'tcp://127.0.0.1:8529', database: '_system', username: 'root'
```
As before, with `--http.keep-alive-timeout 1` the connection succeeds.

**Comment by graetzer** [comment]
you can still use the shell just the message is wrong, anyway I fixed it in 3.6 (upcoming) https://github.com/arangodb/arangodb/pull/10530

**Comment by zickgraf** [comment]
Thanks! I could imagine this also fixes our original problems with Apache. I will retest this once the PR is merged.

**Comment by zickgraf** [comment]
I can confirm that #10530 fixes both the issue described here as well as our original problem with Apache. Thanks! What's the policy for closing issue, shall I close this issue, or will you close it once 3.6 is released?

**Comment by maxkernbach** [comment]
Hi @zickgraf,

We usually close issues once the fixed version has been released.
Fix https://github.com/arangodb/arangodb/pull/10530 is included in 3.6.0, which is available since yesterday - therefore I am closing this issue.

You can find all about 3.6 in our [release blogpost](https://www.arangodb.com/2019/12/arangodb-3-6-release-whats-new/).
